### ISSUE TITLE
"unable to open database file" error from Tx.Exec()

### ISSUE THREAD

**Issue by boramalper** [mattn/go-sqlite3#618]
Hey,

after some time my program has been running, I receive "unable to open database file" error while trying to execute a transaction using `Tx.Exec()`.

I have no clue what might be the cause or even how to debug this. I assume it fails to open a *temporary* file, which I can "fix" by `PRAGMA temp_store=2;` (**edit:** I've tried with `temp_store=2` again but got the same error) but I'd like to know (and report) what might be the issue.

What I know:

- Only **one** goroutine accesses the database *at all times*, and the database is kept open until the program terminates.
- Database URL is:
  
  ```
  /home/(...)/database.sqlite3?_journal_mode=WAL
  ```

- As soon as the database opened, the following statements are executed:

  ```sql
  PRAGMA journal_mode=WAL;
  PRAGMA temp_store=1;
  PRAGMA foreign_keys=ON;
  PRAGMA encoding='UTF-8';
  ```

- Last successful insert was on 2018-08-16T11:29:59.401Z and it failed on 2018-08-16T11:30:04.792Z; the difference in-between is 5 seconds...

- Interestingly, when the error occurred, I had ~1014 TCP connections open, so I thought I might have run out of file descriptors but that can't be the case either:

  ```
  $ cat /proc/sys/fs/file-max
  47804
  ```

- I have (and had) disk space available:

  ```
  $ df -h
  Filesystem      Size  Used Avail Use% Mounted on
  udev            236M     0  236M   0% /dev
  tmpfs            49M  5.5M   44M  12% /run
  /dev/xvda1       16G  5.6G  9.9G  37% /
  tmpfs           244M     0  244M   0% /dev/shm
  tmpfs           5.0M     0  5.0M   0% /run/lock
  tmpfs           244M     0  244M   0% /sys/fs/cgroup
  tmpfs            49M     0   49M   0% /run/user/1000
  ```

- Compilation and versions:

  ```
  $ go version
  go version go1.10 linux/amd64
  $ go build --tags fts5 "-ldflags=-s -w" -o /home/bora/labs/GOPATH/bin/magneticod github.com/boramalper/magnetico/cmd/magneticod
  ```

   ```go
  libVersion: 3.24.0  libVersionNumber: 3024000  sourceID: 2018-06-04 19:24:41 c7ee0833225bfd8c5ec2f9bf62b97c4e04d03bd9566366d5221ac8fb199aalt2  // sqlite3.Version()
   ```

- Database schema:

  ```sql
  $ ./sqlite3 .local/share/magneticod/database.sqlite3 
  SQLite version 3.24.0 2018-06-04 19:24:41
  Enter ".help" for usage hints.
  sqlite> .schema
  CREATE TABLE torrents (
  	id             INTEGER PRIMARY KEY,
  	info_hash      BLOB NOT NULL UNIQUE,
  	name           TEXT NOT NULL,
  	total_size     INTEGER NOT NULL CHECK(total_size > 0),
  	discovered_on  INTEGER NOT NULL CHECK(discovered_on > 0), 
  	updated_on     INTEGER CHECK (updated_on > 0) DEFAULT NULL,
  	n_seeders      INTEGER CHECK ((updated_on IS NOT NULL AND n_seeders >= 0) OR (updated_on IS NULL AND n_seeders IS NULL)) DEFAULT NULL,
  	n_leechers     INTEGER CHECK ((updated_on IS NOT NULL AND n_leechers >= 0) OR (updated_on IS NULL AND n_leechers IS NULL)) DEFAULT NULL,
  	modified_on    INTEGER NOT NULL CHECK (modified_on >= discovered_on AND (updated_on IS NOT NULL OR modified_on >= updated_on)) DEFAULT 32503680000
  );

  CREATE TABLE files (
  	id          INTEGER PRIMARY KEY,
  	torrent_id  INTEGER REFERENCES torrents ON DELETE CASCADE ON UPDATE RESTRICT,
  	size        INTEGER NOT NULL,
  	path        TEXT NOT NULL, 
  	is_readme   INTEGER CHECK (is_readme IS NULL OR is_readme=1) DEFAULT NULL,
  	content     TEXT    CHECK ((content IS NULL AND is_readme IS NULL) OR (content IS NOT NULL AND is_readme=1)) DEFAULT NULL
  );

  CREATE UNIQUE INDEX info_hash_index ON torrents (info_hash);
  CREATE UNIQUE INDEX readme_index ON files (torrent_id, is_readme);
  CREATE VIRTUAL TABLE torrents_idx USING fts5(name, content='torrents', content_rowid='id', tokenize="porter unicode61 separators ' !""#$%&''()*+,-./:;<=>?@[\]^_`{|}~'")
  /* torrents_idx(name) */;
  CREATE TABLE IF NOT EXISTS 'torrents_idx_data'(id INTEGER PRIMARY KEY, block BLOB);
  CREATE TABLE IF NOT EXISTS 'torrents_idx_idx'(segid, term, pgno, PRIMARY KEY(segid, term)) WITHOUT ROWID;
  CREATE TABLE IF NOT EXISTS 'torrents_idx_docsize'(id INTEGER PRIMARY KEY, sz BLOB);
  CREATE TABLE IF NOT EXISTS 'torrents_idx_config'(k PRIMARY KEY, v) WITHOUT ROWID;
  CREATE TRIGGER torrents_idx_ai_t AFTER INSERT ON torrents BEGIN
  	INSERT INTO torrents_idx(rowid, name) VALUES (new.id, new.name);
  END;
  CREATE TRIGGER torrents_idx_ad_t AFTER DELETE ON torrents BEGIN
  	INSERT INTO torrents_idx(torrents_idx, rowid, name) VALUES('delete', old.id, old.name);
  END;
  CREATE TRIGGER torrents_idx_au_t AFTER UPDATE ON torrents BEGIN
  	INSERT INTO torrents_idx(torrents_idx, rowid, name) VALUES('delete', old.id, old.name);
  	INSERT INTO torrents_idx(rowid, name) VALUES (new.id, new.name);
  	END;
  CREATE TRIGGER "torrents_modified_on_default_t" AFTER INSERT ON "torrents" BEGIN
  	UPDATE "torrents" SET "modified_on" = NEW."discovered_on" WHERE "id" = NEW."id" AND NEW."modified_on" = 32503680000;
  END;
  CREATE INDEX modified_on_index ON torrents (modified_on);
  ```

Thanks!

**Comment by rittneje** [comment]
Not sure if this would be contributing to your issue, but one of the problems we've observed in the past is the standard library's attempt to pool connections. (This makes more sense for database connections that are actual network connections, as opposed to SQLite.) Take a look at:

- [SetConnMaxLifetime](https://golang.org/pkg/database/sql/#DB.SetConnMaxLifetime)
- [SetMaxIdleConns](https://golang.org/pkg/database/sql/#DB.SetMaxIdleConns)
- [SetMaxOpenConns](https://golang.org/pkg/database/sql/#DB.SetMaxOpenConns)

Basically, the problem we encountered was that most pragmas (except specifically `PRAGMA journal_mode=WAL`, as per the [documentation](https://www.sqlite.org/pragma.html#pragma_journal_mode)) apply to the _connection_, so if the standard library is opening/closing connections behind your back for pooling purposes, it can lead to unintended behavior.

**Comment by boramalper** [comment]
> - Interestingly, when the error occurred, I had ~1014 TCP connections open, so I thought I might have run out of file descriptors but that can't be the case either:
>
> ```bash
> $ cat /proc/sys/fs/file-max
> 47804
> ```

Apparently the case was I ran out of file descriptors. =) Max number of open files is/was indeed 1,024 -- which could be found by `cat /proc/<PID>/limits`.

@rittneje's answer also helped, since I was able to limit the (max) number of open connections, instead of the default unlimited.

----

I am leaving the issue open since I believe go-sqlite3 should report the cause of error (too many open files) rather than returning "unable to open database file" only.

**Comment by rittneje** [comment]
My assumption is that the error in question is being returned by SQLite itself - see [`SQLITE_CANTOPEN`](https://www.sqlite.org/c3ref/c_abort.html). You might want to ask the SQLite maintainers to add an [extended result code](https://www.sqlite.org/c3ref/c_abort_rollback.html) for this situation.

**Comment by boramalper** [comment]
I see. I’m closing the issue then, feel free to contact SQLite maintainers if you wish, but I’m too lazy for that. =)

**Comment by rittneje** [comment]
@boramalper I just found [this function](https://www.sqlite.org/c3ref/system_errno.html), which may be of use to you. I'm not entirely certain, but I would guess that the errno would be `EMFILE` or `ENFILE` in this case (according to the [man page](http://man7.org/linux/man-pages/man2/open.2.html)). So perhaps the result of `system_errno` could be added to the `Error` struct?

**Comment by boramalper** [comment]
@rittneje I think that's a great idea, I've created a new issue. =)
